### PR TITLE
Add stream layer read example.

### PR DESCRIPTION
**Pull request by LiubovDidkivska**
Create stream layer client with user specified catalog, layer id and credentials.
Try to poll data till all messages are consumed.
Check received messages from a stream layer. If received data is more than 1mb and data handle is returned, call GetData for each received message.

Relates_to:OLPEDGE-1306

Signed-off-by: Liubov Didkivska <ext-liubov.didkivska@here.com>

### CODE REVIEW

File: examples/StreamLayerReadExample.cpp [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    int res = 0;
+    if (CreateSubscription(client, subscribe_request) != 0 ||
+        RunPoll(client) != 0) {
+      res = -1;
+    }
+    if (DeleteSubscription(client) != 0) {
+      res = -1;
+    }
+    return res;
+  };
```
Comment (andescu) [inline examples/StreamLayerReadExample.cpp]: This looks very cryptic. Is there no better way to do it? Looking at this code I really don't understand how to use the stream layer client to poll data. Please try to keep the overhead, e.g. logs and return values as slim as possible. You can simply return bool from functions that only use 0 or -1.

Here is an example from the Kafka CPP client -> https://github.com/mfontanini/cppkafka/blob/master/examples/consumer_example.cpp

File: examples/StreamLayerReadExample.cpp [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    RunPoll(client);
+    if (!DeleteSubscription(client)) {
+      return false;
+    }
+    return true;
+  };
+
+  int result = 0;
+  if (subscription_mode ==
+      olp::dataservice::read::SubscribeRequest::SubscriptionMode::kParallel) {
```
Comment (andescu) [inline examples/StreamLayerReadExample.cpp]: For the sake of readability maybe add an alias for the namespace olp::dataservice::read or for the SubscribeRequest.

File: examples/StreamLayerReadExample.cpp [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    // With parallel subscription you can read large volumes of data in a
+    // parallel manner. The subscription and message reading workflow is similar
+    // to serial subscription except that you are allowed to create multiple
+    // subscriptions for the same aid, catalog, layer, group.id combination
```
Comment (andescu) [inline examples/StreamLayerReadExample.cpp]:
```suggestion
    // subscriptions for the same HRN, layer and group.id combination
```

File: examples/StreamLayerReadExample.cpp [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    // using multiple processes/threads. This allows you to read and commit
+    // messages for each subscription in parallel.
+    OLP_SDK_LOG_INFO_F(kLogTag, "Run Poll in parallel, threads count = %u",
```
Comment (andescu) [inline examples/StreamLayerReadExample.cpp]:
```suggestion
    OLP_SDK_LOG_INFO_F(kLogTag, "Starting parallel subscription mode, threads=%u",
```

File: examples/StreamLayerReadExample.cpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                       kNumberOfThreads);
+    std::atomic<int> value(0);
+    std::thread threads[kNumberOfThreads];
+    for (auto& thread : threads) {
+      thread = std::thread([&]() {
+        if (!read_from_stream_layer()) {
+          // acumulate result from all threads
+          value.store(-1);
+        }
+      });
```
Comment (andescu) [inline examples/StreamLayerReadExample.cpp]: It is very confusing that you use lambda in lambda. Why dont you directly set the atomic in read_from_stream_layer()?

File: examples/StreamLayerReadExample.cpp [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    // With parallel subscription you can read large volumes of data in a
+    // parallel manner. The subscription and message reading workflow is similar
+    // to serial subscription except that you are allowed to create multiple
+    // subscriptions for the same aid, catalog, layer, group.id combination
+    // using multiple processes/threads. This allows you to read and commit
+    // messages for each subscription in parallel.
+    OLP_SDK_LOG_INFO_F(kLogTag, "Run Poll in parallel, threads count = %u",
+                       kNumberOfThreads);
+    std::atomic<int> value(0);
+    std::thread threads[kNumberOfThreads];
```
Comment (andescu) [inline examples/StreamLayerReadExample.cpp]: Please use vector and emplace back.

File: examples/StreamLayerReadExample.cpp [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    for (auto& thread : threads) {
+      thread.join();
+    }
+    result = value.load();
+  } else {
+    // With serial subscription you can read  smaller volumes of data with a
+    // single subscription.
+    result = read_from_stream_layer() ? 0 : -1;
+  }
+  return result;
```
Comment (andescu) [inline examples/StreamLayerReadExample.cpp]: Again here I would revert the logic:

```cpp
if (subscription_mode == kSerial) {
  return read_from_stream_layer() ? 0 : -1;
}

// With parallel subscription you can read large volumes of data in a
// parallel manner. The subscription and message reading workflow is similar
// to serial subscription except that you are allowed to create multiple
// subscriptions for the same aid, catalog, layer, group.id combination
.....

return value.load();
```

File: examples/CMakeLists.txt [hunk 8]
```diff
@@ -51,6 +52,15 @@ else()
     target_link_libraries(${OLP_SDK_DATASERVICE_READ_EXAMPLE_TARGET}
         olp-cpp-sdk-authentication
         olp-cpp-sdk-dataservice-read)
+        
+    add_library(${OLP_SDK_DATASERVICE_READ_STREAM_LAYER_EXAMPLE_TARGET}
+        ${CMAKE_CURRENT_SOURCE_DIR}/StreamLayerReadExample.cpp
+        ${CMAKE_CURRENT_SOURCE_DIR}/Examples.h
+        ${CMAKE_CURRENT_SOURCE_DIR}/StreamLayerReadExample.h)
```
Comment (meravingen123) [inline examples/CMakeLists.txt]: Why CMAKE_CURRENT_SOURCE_DIR is needed?

File: examples/main.cpp [hunk 9]
```diff
@@ -35,20 +36,26 @@ enum Examples : int {
...
+    "to"
     "get data(used in read example, optional). \n"
-    " -l [--layer_id] the layer ID inside the catalog where you want to "
+    " -l, --layer-id \n\tThe layer ID inside the catalog where you want to "
     "publish data to(required for write example). \n"
-    " -h [--help]: show usage \n For "
-    "instructions on how to get the access key ID and access key secret, see "
+    " -t,--type-of-subscription[=serial|parallel] \n\tType of subscription  "
+    "(used for read_stream test). The default option is serial. \n"
+    " -h,--help \n\tShow usage \nFor instructions on how to get the access key "
```
Comment (andescu) [inline examples/main.cpp]:
```suggestion
    " -h, --help \n\tShow usage \nFor instructions on how to get the access key "
```

File: examples/main.cpp [hunk 10]
```diff
@@ -35,20 +36,26 @@ enum Examples : int {
...
+    " -c, --catalog \n\tCatalog HRN (HERE Resource Name). \n"
+    " -v,--catalog-version \n\tThe version of the catalog from which you want "
+    "to"
     "get data(used in read example, optional). \n"
-    " -l [--layer_id] the layer ID inside the catalog where you want to "
+    " -l, --layer-id \n\tThe layer ID inside the catalog where you want to "
     "publish data to(required for write example). \n"
-    " -h [--help]: show usage \n For "
-    "instructions on how to get the access key ID and access key secret, see "
+    " -t,--type-of-subscription[=serial|parallel] \n\tType of subscription  "
```
Comment (andescu) [inline examples/main.cpp]:
```suggestion
    " -t, --type-of-subscription[=serial|parallel] \n\tType of subscription  "
```

File: examples/main.cpp [hunk 11]
```diff
@@ -35,20 +36,26 @@ enum Examples : int {
...
-    "example [read, write, cache] \n -i [--key_id] here.access.key.id \n -s "
-    "[--key_secret] here.access.key.secret \n"
-    " -c [--catalog] catalog HRN (HERE Resource Name). \n"
-    " -v [--catalog_version] The version of the catalog from which you wan to"
+    "usuage is \n -a,--all : run all examples \n "
+    "-e, --example[=read|read_stream|write|cache]  \n\tRun "
+    "example\n -i,--key-id \n\there.access.key.id \n -s, --key-secret "
+    "\n\there.access.key.secret \n"
+    " -c, --catalog \n\tCatalog HRN (HERE Resource Name). \n"
+    " -v,--catalog-version \n\tThe version of the catalog from which you want "
```
Comment (andescu) [inline examples/main.cpp]:
```suggestion
    " -v, --catalog-version \n\tThe version of the catalog from which you want "
```

File: examples/StreamLayerReadExample.cpp [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+    // Create stream layer client with settings and catalog, layer specified
+    StreamLayerClient client(olp::client::HRN{catalog}, layer_id,
+                             client_settings);
+    if (!CreateSubscription(client, subscribe_request)) {
+      value.store(-1);
+    }
+    RunPoll(client);
+    if (!DeleteSubscription(client)) {
+      value.store(-1);
+    }
```
Comment (andescu) [inline examples/StreamLayerReadExample.cpp]:
```suggestion
    if (!CreateSubscription(client, subscribe_request)) {
      value.store(-1);
    }
    
    RunPoll(client);
    
    if (!DeleteSubscription(client)) {
      value.store(-1);
    }
```

**Pull request merged**